### DISCUSSION TITLE
Webpack 5 performance issue with PouchDB

### DISCUSSION THREAD

**Discussion by TheSlimvReal** [webpack/webpack#13799] [Q&A]
I recently upgraded my Angular application to Webpack 5. I previously patched the webpack config to polyfill node packages with `node: {crypto: true, stream: true, fs: 'empty'}`. For the migration to Webpack 5 I changed it to

```
promise.fallback = {
   crypto: require.resolve('crypto-browserify'),
   stream: require.resolve('stream-browserify'),
   path: require.resolve('path-browserify'),
}
```

Additionally, I suddenly had to manually create a polyfill for `Buffer`
```
import * as buffer from "buffer";
(window as any).Buffer = buffer.Buffer;
```

Since I applied these changes, the PouchDB I am using became incredibly slow. Especially the creation of indices and the querying of these indices takes really really long. 

Do you have any ideas where this problem is coming from or how it can be fixed?

**Comment by alexander-akait** [comment]
hm, need to profile, hard to say, webpack include the same polyfill before, maybe some of them was updated since webpack v4 and get perf regressions...

**Reply by TheSlimvReal** [reply]
I came across the same performance issues in Webpack 4. when I replace 

```node: {crypto: true, stream: true, fs: 'empty'}``` 

with 

```
resolve.alias: {
    crypto: require.resolve('crypto-browserify'),
    path: require.resolve('path-browserify'),
    stream: require.resolve('stream-browserify'),
}
``` 
(I used `resolve.alias` because `resolve.fallback` was only added in Webpack 5)

Is there any difference between the two approaches?

**Reply by alexander-akait** [reply]
as I said above it means webpck v4 uses old version of `crypto-browserify`/`path-browserify`/`stream-browserify`, so they don't have perf problems, please check their version in webpack v4 and in your repo

**Reply by TheSlimvReal** [reply]
In my `package.json` I only have one version of each of each of the packages, so I guess it will use the same polyfills in both cases. Is there a way do check at runtime which package and version is used for e.g. `crypto` or `stream`?

Also, when I do the manual Polyfill (using `resolve.alias` or `resolve.fallback`) I have to polyfill the global `Buffer` variable manually:

```
import * as buffer from "buffer";
(window as any).Buffer = buffer.Buffer;
```

**Reply by alexander-akait** [reply]
in this case we need profile code, I can't say why it slowly...

**Reply by TheSlimvReal** [reply]
How exactly should this profile code look? I encountered the problem at [my public repository](https://github.com/Aam-Digital/ndb-core) which is a rather complex web-app but the actual performance issue comes down to the lines I mentioned in this discussion earlier. I am happy to talk you through the changes that have to be made in order to experience the performance issue, but I don't know how to profile the code myself.

**Reply by alexander-akait** [reply]
Exactly how you do it for other code, https://nodejs.org/en/docs/guides/simple-profiling/
